fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Compensation.java

```diff
@@ -395,13 +402,13 @@ WithSelectCompensation derivedWithPredicateCompensationMap(@Nonnull Compensation
         @Override
         default Compensation union(@Nonnull Compensation otherCompensation) {
             if (!(otherCompensation instanceof WithSelectCompensation)) {
-                return Compensation.super.union(otherCompensation);
+                return otherCompensation.union(this);
```

Comment:
Trying to explain the naming is kind of making me realize that the naming may be a misnomer. We can rename this but this is how it was dreamt up in the first place.

**INTERSECT**
You have a bunch of compensations and you want to intersect the records coming from the index.

Example:
```
table T: a, b, c
INDEX: i1: (a, b)
INDEX: i2: (a, c)

SELECT *
FROM T
WHERE a = 3 AND b = 4 AND c > 5
``` 

If I didn't make a mistake (:-)), this should give you an intersection plan of `i1` and `i2` as `b` is equality bound and both scans give you `a, c` order.

**UNION**
You have a bunch of compensations for the current e.g. `SelectExpression` that both have children compensations that are _needed_ (i.e. they are non-trivial). Before you can somehow apply the compensations for the current expression, you need to take care of the compensations of your children. Expressions form a dag, matches form a tree, compensations form a list. Now most of the time, having a non-trivial compensation in the children, prevents the parent match to never actually be formed in the first place but it could happen. In this case, you can apply one first then the other. Or, if you know both are in form of a `ForMatch` compensation, you can just union the predicates you need to reapply together.

Again, I need to admit that the name-giving is not very intuitive. If you file a ticket, we can change that in a separate PR.